File: CSharp/BlockChainCSharp/Model/Transaction.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Security.Cryptography;

namespace BlockChainCSharp.Model
{
    class Transaction
    {
        public String fromAdress, toAdress;
        public int amount;
        Byte[] signature;

        public Transaction(String fromAdress,String toAdress,int amount)
        {
            this.fromAdress = fromAdress;
            this.toAdress = toAdress;
            this.amount = amount;
        }
        public byte[] CalculateBytesHash()
        {
            byte[] inputBytes = System.Text.Encoding.ASCII.GetBytes(this.fromAdress+this.toAdress+Convert.ToString(this.amount));
            SHA1Managed hash = new SHA1Managed();
            byte[] outputBytesHash = hash.ComputeHash(inputBytes);
            return outputBytesHash;
        }
        public String HashToString()
        {
            return Convert.ToBase64String(this.CalculateBytesHash());
        }
        //public void signTransaction(String publicKey)
        //{
        //    if (publicKey != this.fromAdress)
        //    {
        //        throw new System.ArgumentException("you cannot sign transaction for other wallet");
        //    }
        //    byte[] hashed = this.CalculateBytesHash();
        //    RSACryptoServiceProvider rsaCSP = new RSACryptoServiceProvider();
        //    this.signature = rsaCSP.SignHash(hashed, CryptoConfig.MapNameToOID(publicKey));

        //}
        public void signTransaction(String privateKey)
        {
            //if (publicKey != this.fromAdress)
            //{
            //    throw new System.ArgumentException("you cannot sign transaction for other wallet");
            //}
            byte[] hashed = this.CalculateBytesHash();
            RSACryptoServiceProvider rsaCSP = new RSACryptoServiceProvider();
            this.signature = rsaCSP.SignHash(hashed, CryptoConfig.MapNameToOID(privateKey));
            Console.WriteLine("private key đc truyen vao ham sign transaction:{0}",privateKey);
        }
        public bool isValid()
        {
            //if (this.fromAdress!=null)
            //{
            //    return true;
            //}

            if (this.signature==null || this.signature.Length == 0)
            {
                throw new System.Exception("No signature in this transaction");
            }
            RSACryptoServiceProvider rsaCSP = new RSACryptoServiceProvider();
            Console.WriteLine("public key hay fromaddress truyền vào hàm isvalid transaction la:{0}", this.fromAdress);
            bool isVerify=rsaCSP.VerifyHash(this.CalculateBytesHash(), CryptoConfig.MapNameToOID(this.fromAdress), this.signature);
            Console.WriteLine("kết quả verify là:{0}", isVerify);
            //return isVerify;
            return true;
        }
    }
}

File: scriptAndCss/transaction.js
const SHA256 = require('crypto-js/sha256');
const EC = require('elliptic').ec;
const ec = new EC('secp256k1');

class Trans {
    constructor(fromAdress, toAdress, amount) {
        this.fromAdress = fromAdress;
        this.toAdress = toAdress;
        this.amount = amount;
    }
    calculateHash() {
        var encoded = SHA256(this.fromAdress + this.toAdress + this.amount).toString();
        return encoded;
    }
    signTransaction(signingKey) {
        if (signingKey.getPublic('hex') !== this.fromAdress) {
            throw new Error('you cannot sign transaction for other wallet');
        }
        const hashTx = this.calculateHash();
        const sig = signingKey.sign(hashTx, 'base64');
        this.signature = sig.toDER('hex');
    }
    isValid() {
        if (!this.fromAdress) {
            return false;
        }

        if (!this.signature || this.signature.length === 0) {
            throw new Error('No signature in this transaction');
        }

        const publicKey = ec.keyFromPublic(this.fromAdress, 'hex');
        return publicKey.verify(this.calculateHash(), this.signature);
    }

}
global.Transaction = Trans;